AX: Merge AccessibilityImageMapLink into AccessibilityNodeObject

hoffmanjoshua opened the pull request:
#### 9c4a4858882d8ef4d5e7c957519b95bbf511b1a1
<pre>
AX: Merge AccessibilityImageMapLink into AccessibilityNodeObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=298584">https://bugs.webkit.org/show_bug.cgi?id=298584</a>
<a href="https://rdar.apple.com/160171945">rdar://160171945</a>

Reviewed by Tyler Wilcock.

Merge all of the logic from AccessibilityImageMapLink into AccessibilityNodeObject.

Critically, this adds a way to create AccessibilityNodeObjects explicitly. This will be useful
in our end goal of collapsing all node-object subclasses into AccessibilityNodeObject.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isImageMapLink const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXImageMapHelpers.cpp: Added.
(WebCore::AXImageMapHelpers::accessibilityText):
(WebCore::AXImageMapHelpers::rendererFromAreaElement):
* Source/WebCore/accessibility/AXImageMapHelpers.h: Added.
* Source/WebCore/accessibility/AXLoggerBase.cpp:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::focusedImageMapUIElement):
(WebCore::AXObjectCache::createFromNode):
* Source/WebCore/accessibility/AXTreeStore.cpp:
* Source/WebCore/accessibility/AccessibilityImageMapLink.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityImageMapLink.h: Removed.
* Source/WebCore/accessibility/AccessibilityListBoxOption.cpp:
* Source/WebCore/accessibility/AccessibilityMenuList.cpp:
* Source/WebCore/accessibility/AccessibilityMenuListPopup.cpp:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::create):
(WebCore::AccessibilityNodeObject::parentObject const):
(WebCore::AccessibilityNodeObject::elementRect const):
(WebCore::AccessibilityNodeObject::elementPath const):
(WebCore::AccessibilityNodeObject::determineAccessibilityRole):
(WebCore::AccessibilityNodeObject::computeIsIgnored const):
(WebCore::AccessibilityNodeObject::isEnabled const):
(WebCore::AccessibilityNodeObject::anchorElement const):
(WebCore::AccessibilityNodeObject::usesAltForTextComputation const):
(WebCore::AccessibilityNodeObject::accessibilityText const):
(WebCore::AccessibilityNodeObject::description const):
(WebCore::AccessibilityNodeObject::url const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isImageMapLink const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::documentLinks):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::createIsolatedObjectData):

Canonical link: <a href="https://commits.webkit.org/299788@main">https://commits.webkit.org/299788@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe9b8f79b67bd84fa4a92e01a926ea9646fa42dc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120133 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39827 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30478 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126481 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72206 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f33b56a8-a158-4b4b-abca-9eecbca57d9d) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122009 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40523 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48404 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91235 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60546 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a7e7142d-4cc5-48ae-8547-7107485395d4) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123085 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32374 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107711 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71786 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/57624a49-9c13-4db0-a3d2-17b8f8e67962) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31406 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25817 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/70953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101853 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26004 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129386 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47053 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35694 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99854 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47419 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103899 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99697 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25320 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45152 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23173 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43684 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46915 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52621 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46383 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49730 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48067 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXImageMapHelpers.cpp:
```diff
@@ -0,0 +1,60 @@
+
+using namespace HTMLNames;
+
+void AXImageMapHelpers::accessibilityText(const HTMLAreaElement& attachmentElement, String description, Vector<AccessibilityText>& textOrder)
```
`attachmentElement` may be a confusing name for an `HTMLAreaElement`, considering `HTMLAttachmentElement` exists

twilco replied:
This method can take an r-value reference for description to avoid a copy

```cpp
String&& description
```

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -413,6 +444,9 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRole()
     if (isExposedTableRow())
         return AccessibilityRole::Row;
 
+    if (isImageMapLink())
+        return !url().isEmpty() ? AccessibilityRole::Link : AccessibilityRole::Generic;
```
Do you have any idea if we re-compute object's roles when their URL change, considering role seems dependent on that state? If not, maybe we can add a FIXME here to clean up later?

hoffmanjoshua replied:
We don't... we just update the link and URL properties. I'll add this fixme!

twilco replied:
Thank you 🙏

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -919,6 +953,9 @@ bool AccessibilityNodeObject::computeIsIgnored() const
         return true;
 
+    if (isImageMapLink())
+        return false;
```
So we unignore image-map links even if they are generic (e.g. have no URL)? I wonder if that's the best behavior (maybe?)

hoffmanjoshua replied:
I wonder if this is because people are using image map links to label parts of an image? Not a correct use, but one use case I could see

twilco replied:
OK fair, well you're just moving this logic so let's leave it

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -1662,6 +1662,12 @@ AXCoreObject* AXIsolatedObject::parentTableIfExposedTableRow() const
+{
+    ASSERT_NOT_REACHED();
+    return false;
+}
```
Does `isImageMapLink` really need to be virtual on `AXCoreObject`? Ideally we wouldn't have many or any of these tens of hardcoded `ASSERT_NOT_REACHED();` isolated object overrides — I spent a long time a while ago trying to remove them 🙂

hoffmanjoshua replied:
It's tricky, because anywhere we get an AccessibilityChildrenVector, we would have to downcast the result just to check if something is an image map link, which also has a cost. Not sure which is more ideal here?

twilco replied:
I think we're paying that cost either way by increasing the virtuality of the function (the compiler just makes the downcast "invisible" / handled by the vtable). Couldn't we just make this function totally non-virtual on `AXCoreObject`, and have the implementation do this? This actually removes one virtual function call from the chain (only `element()`), and allows us to avoid adding this `AXIsolatedObject::isImageMapLink()` override, and avoids having to add `downcasts` in more places

```cpp
AXCoreObject::isImageMapLink() const
{
    RefPtr element = this->element();
    return element && is<HTMLAreaElement>(*element);
}
```

hoffmanjoshua replied:
great call, updated!

hoffmanjoshua committed:
AX: Merge AccessibilityImageMapLink into AccessibilityNodeObject
https://bugs.webkit.org/show_bug.cgi?id=298584
rdar://160171945

Reviewed by Tyler Wilcock.

Merge all of the logic from AccessibilityImageMapLink into AccessibilityNodeObject.

Critically, this adds a way to create AccessibilityNodeObjects explicitly. This will be useful
in our end goal of collapsing all node-object subclasses into AccessibilityNodeObject.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isImageMapLink const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXImageMapHelpers.cpp: Added.
(WebCore::AXImageMapHelpers::accessibilityText):
(WebCore::AXImageMapHelpers::rendererFromAreaElement):
* Source/WebCore/accessibility/AXImageMapHelpers.h: Added.
* Source/WebCore/accessibility/AXLoggerBase.cpp:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::focusedImageMapUIElement):
(WebCore::AXObjectCache::createFromNode):
* Source/WebCore/accessibility/AXTreeStore.cpp:
* Source/WebCore/accessibility/AccessibilityImageMapLink.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityImageMapLink.h: Removed.
* Source/WebCore/accessibility/AccessibilityListBoxOption.cpp:
* Source/WebCore/accessibility/AccessibilityMenuList.cpp:
* Source/WebCore/accessibility/AccessibilityMenuListPopup.cpp:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::create):
(WebCore::AccessibilityNodeObject::parentObject const):
(WebCore::AccessibilityNodeObject::elementRect const):
(WebCore::AccessibilityNodeObject::elementPath const):
(WebCore::AccessibilityNodeObject::determineAccessibilityRole):
(WebCore::AccessibilityNodeObject::computeIsIgnored const):
(WebCore::AccessibilityNodeObject::isEnabled const):
(WebCore::AccessibilityNodeObject::anchorElement const):
(WebCore::AccessibilityNodeObject::usesAltForTextComputation const):
(WebCore::AccessibilityNodeObject::accessibilityText const):
(WebCore::AccessibilityNodeObject::description const):
(WebCore::AccessibilityNodeObject::url const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isImageMapLink const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::documentLinks):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::createIsolatedObjectData):

Canonical link: https://commits.webkit.org/299788@main
